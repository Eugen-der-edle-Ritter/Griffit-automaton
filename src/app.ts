import '../public/style.css';
import { cell } from './modules/types';
import { randomFill, transitionFill } from './modules/filling';

/* INITILIZE AUTOMATA STATE */
const statesArr: string[] = [
  '#113',
  '#014',
  '#b16',
  '#54f',
  '#74f',
  '#fff',
  '#add',
  '#4fc',
  '#185',
  '#8c4',
  '#ef6',
  '#4bd',
  '#27f',
  '#33f',
  '#059',
  '#567',
];
const states: number = statesArr.length;

/* INTERFACE FOR GLOBAL CANVAS AUTOMATON CELL */
interface ICell {
  width: number;
  height: number;
}

window.onload = () => {
  /* GET CANVAS CONTEXT */
  const canvas: HTMLCanvasElement = document.getElementById(
    'canvas'
  ) as HTMLCanvasElement;
  const context: CanvasRenderingContext2D | null = canvas.getContext('2d');

  const { width, height }: ICell = canvas;

  /* DRAW EMPTY RECTANGLE */
  context!.fillRect(0, 0, width, height);

  /* SUBSCRIBE ON TRANSITION-TYPE TOGGLER */
  const ruleSelector: HTMLElement | null =
    document.querySelector('.ruleSelector');
  if (ruleSelector) {
    ruleSelector.addEventListener('change', (e): void => {
      const newRule: string = (e.target as HTMLTextAreaElement).value;
      clearInterval(timer);
      field = randomFill(byteArr.length, height, width);
      timer = drawAutomata(newRule);
    });
  }

  /* TRANSFORM HEX STRING INTO BYTE ARRAYS */
  const byteArr: number[][] = colorTransform(statesArr);

  console.log(byteArr, 'hi')

  /* DEFINE AND FILL 2d ARRAY OF CELL STATES BY RANDOM VALUES */
  let field: cell[][] = randomFill(byteArr.length, height, width);

  /* TRANSFORMING CANVAS CELL'S STATES THROUGH THE TIME */
  let timer = drawAutomata('hash');

  function drawAutomata(rule: string) {
    const timer = setInterval(() => {
      field = transitionFill(rule, field, states, height, width);
      draw();
    }, 50);
    return timer;
  }

  /* CONVERTING HEX INTO BYTE ARRAY */
  function colorTransform(colorArr: string[]): number[][] {
    return colorArr.map((el, i) => {
      return [
        parseInt(colorArr[i].charAt(1) + colorArr[i].charAt(1), 16),
        parseInt(colorArr[i].charAt(2) + colorArr[i].charAt(2), 16),
        parseInt(colorArr[i].charAt(3) + colorArr[i].charAt(3), 16),
      ];
    });
  }

  /* RENDERING CANVAS GRID */
  function draw(): void {
    const imageData = context!.getImageData(0, 0, width, height);
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const offset = (width * y + x) * 4;
        imageData.data[offset] = byteArr[field[y][x]][0];
        imageData.data[offset + 1] = byteArr[field[y][x]][1];
        imageData.data[offset + 2] = byteArr[field[y][x]][2];
      }
    }
    context!.putImageData(imageData, 0, 0);
  }
};
