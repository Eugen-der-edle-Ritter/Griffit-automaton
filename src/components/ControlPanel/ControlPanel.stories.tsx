import React from 'react';
import { Story, Meta } from '@storybook/react';

import { ControlPanel, ControlPanelProps } from './ControlPanel';

export default {
  component: ControlPanel,
  title: 'Control Panel/ControlPanel',
} as Meta;

const Template: Story<ControlPanelProps> = (args) => <ControlPanel {...args} />;

export const PanelExample = Template.bind({});

PanelExample.args = {};
