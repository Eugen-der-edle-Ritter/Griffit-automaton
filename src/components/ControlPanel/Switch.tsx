import React, { FC, ChangeEvent } from 'react';
import styled from '@emotion/styled';

export interface SwitchProps {
  label: string;
  options: string[];
  onChange: (event: ChangeEvent) => void;
}

export const Switch: FC<SwitchProps> = ({ label, options, onChange }) => (
  <Div>
    <Label>{label}: </Label>
    <Select onChange={onChange}>
      {options.map((option, index) => (
        <Option key={index}>{option}</Option>
      ))}
    </Select>
  </Div>
);

const Div = styled.div``;
const Label = styled.label`
  margin-right: 0.5rem;
  color: #e6e6e6;
  font-size: 0.85rem;
`;
const Select = styled.select`
  height: 1.6rem;
  padding-left: 0.2rem;
  color: #dddddd;
  background-color: #383b40;
  border: 0.5px solid #010101;
  border-radius: 0.5rem;
  outline: none;
  cursor: pointer;
`;
const Option = styled.option``;
