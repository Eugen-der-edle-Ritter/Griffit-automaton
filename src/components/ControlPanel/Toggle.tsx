import React, { FC } from 'react';
import styled from '@emotion/styled';

export interface ToggleProps {
  label: string;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Toggle: FC<ToggleProps> = ({ children, onClick }) => (
  <Button onClick={onClick}>{children}</Button>
);

const Button = styled.button`
  margin: 0 0.3rem;
  padding: 0.3rem 0.7rem;
  color: #e6e6e6;
  background-color: #383b40;
  border-radius: 0.5rem;
  border: 0.5px solid #010101;
  font-family: 'Roboto', sans-serif;
  cursor: pointer;
  &:hover {
    background-color: #383b405e;
  }
`;
