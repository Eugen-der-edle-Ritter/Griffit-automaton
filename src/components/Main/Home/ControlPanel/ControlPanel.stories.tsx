import React from 'react';
import { store } from '@/automaton/automaton';
import { Provider } from 'react-redux';

import { Story, Meta } from '@storybook/react';

import { ControlPanel } from './ControlPanel';

export default {
  component: ControlPanel,
  title: 'Control Panel/ControlPanel',
} as Meta;

const Template: Story = (args) => (
  <Provider store={store}>
    <ControlPanel {...args} />
  </Provider>
);

export const PanelExample = Template.bind({});

PanelExample.args = {};
