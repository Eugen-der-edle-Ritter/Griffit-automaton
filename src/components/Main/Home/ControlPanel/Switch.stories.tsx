import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Switch, SwitchProps } from './Switch';

export default {
  component: Switch,
  title: 'Control Panel/Select',
} as Meta;

const Template: Story<SwitchProps> = (args) => <Switch {...args} />;

export const SelectExample = Template.bind({});

SelectExample.args = {
  label: 'Label',
  options: ['option1', 'option2', 'option3'],
};
