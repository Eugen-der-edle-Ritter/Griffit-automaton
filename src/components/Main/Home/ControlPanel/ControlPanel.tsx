import React, { FC, ChangeEvent } from 'react';
import styled from '@emotion/styled';

import { Toggle } from './Toggle';
import { Switch } from './Switch';

export interface ControlPanelProps {
  handleStop: () => void;
  handleStart: () => void;
  handleRule: (event: ChangeEvent) => void;
}

export const ControlPanel: FC<ControlPanelProps> = ({
  handleStop,
  handleStart,
  handleRule,
}) => (
  <Section>
    <Wrapper>
      <Toggle label="start" onClick={handleStop}>
        Stop
      </Toggle>
      <Toggle label="start" onClick={handleStart}>
        Start
      </Toggle>
    </Wrapper>
    <Switch
      label="Rule"
      options={['Hash', 'Demons', 'Venus']}
      onChange={handleRule}
    />
  </Section>
);

const Section = styled.section`
  margin: 1rem auto 0;
  padding: 1rem;
  box-sizing: border-box;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #1b1d20;
  border-radius: 0.7rem 0.7rem 0 0;
`;
const Wrapper = styled.div``;
