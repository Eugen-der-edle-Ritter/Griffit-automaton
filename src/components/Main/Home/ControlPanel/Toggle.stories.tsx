import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Toggle, ToggleProps } from './Toggle';

export default {
  component: Toggle,
  title: 'Control Panel/Button',
} as Meta;

const Template: Story<ToggleProps> = (args) => <Toggle {...args} />;

export const ButtonExample = Template.bind({});

ButtonExample.args = {
  label: 'Example',
};
