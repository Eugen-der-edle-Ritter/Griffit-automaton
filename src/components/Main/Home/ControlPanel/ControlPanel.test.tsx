import React from 'react';
import { Provider, useDispatch, useSelector } from 'react-redux';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { store } from '@/app/store';

import { ControlPanel } from './ControlPanel';

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest
    .fn()
    .mockReturnValueOnce({ past: [], future: [] })
    .mockReturnValueOnce({ isPlaying: true }),
}));

describe('ControlPanel test cases', () => {
  it('ControlPanel dispatching', () => {
    const mockDispatch = jest.fn();

    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);

    const { asFragment } = render(
      <Provider store={store}>
        <ControlPanel />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();

    const buttons = screen.getAllByRole('button');
    buttons.forEach((el) => userEvent.click(el));

    expect(mockDispatch).toHaveBeenCalledTimes(2);
  });
});
