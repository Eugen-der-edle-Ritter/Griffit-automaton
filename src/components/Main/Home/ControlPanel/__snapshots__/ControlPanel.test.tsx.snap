// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel test cases ControlPanel renders on pause 1`] = `
<DocumentFragment>
  <section
    class="css-18a9xii"
  >
    <div
      class="css-0"
    >
      <button
        class="css-148csyh"
        data-testid="undo"
      >
        <span
          aria-label="step-backward"
          class="anticon anticon-step-backward"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="step-backward"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M347.6 528.95l383.2 301.02c14.25 11.2 35.2 1.1 35.2-16.95V210.97c0-18.05-20.95-28.14-35.2-16.94L347.6 495.05a21.53 21.53 0 000 33.9M330 864h-64a8 8 0 01-8-8V168a8 8 0 018-8h64a8 8 0 018 8v688a8 8 0 01-8 8"
            />
          </svg>
        </span>
      </button>
      <button
        class="css-2ly8m7"
      >
        <span
          aria-label="caret-right"
          class="anticon anticon-caret-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="caret-right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M715.8 493.5L335 165.1c-14.2-12.2-35-1.2-35 18.5v656.8c0 19.7 20.8 30.7 35 18.5l380.8-328.4c10.9-9.4 10.9-27.6 0-37z"
            />
          </svg>
        </span>
      </button>
      <button
        class="css-148csyh"
        data-testid="redo"
      >
        <span
          aria-label="step-forward"
          class="anticon anticon-step-forward"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="step-forward"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M676.4 528.95L293.2 829.97c-14.25 11.2-35.2 1.1-35.2-16.95V210.97c0-18.05 20.95-28.14 35.2-16.94l383.2 301.02a21.53 21.53 0 010 33.9M694 864h64a8 8 0 008-8V168a8 8 0 00-8-8h-64a8 8 0 00-8 8v688a8 8 0 008 8"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="css-0"
    >
      <label
        class="css-jm2r67"
      >
        Rule: 
      </label>
      <select
        class="css-vqx920"
        data-testid="select"
        disabled=""
      >
        <option
          class="css-0"
          value="Hash"
        >
          Hash
        </option>
        <option
          class="css-0"
          value="Demons"
        >
          Demons
        </option>
        <option
          class="css-0"
          value="Venus"
        >
          Venus
        </option>
      </select>
    </div>
    <button
      class="css-2ly8m7"
    >
      <span
        aria-label="bg-colors"
        class="anticon anticon-bg-colors"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="bg-colors"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M766.4 744.3c43.7 0 79.4-36.2 79.4-80.5 0-53.5-79.4-140.8-79.4-140.8S687 610.3 687 663.8c0 44.3 35.7 80.5 79.4 80.5zm-377.1-44.1c7.1 7.1 18.6 7.1 25.6 0l256.1-256c7.1-7.1 7.1-18.6 0-25.6l-256-256c-.6-.6-1.3-1.2-2-1.7l-78.2-78.2a9.11 9.11 0 00-12.8 0l-48 48a9.11 9.11 0 000 12.8l67.2 67.2-207.8 207.9c-7.1 7.1-7.1 18.6 0 25.6l255.9 256zm12.9-448.6l178.9 178.9H223.4l178.8-178.9zM904 816H120c-4.4 0-8 3.6-8 8v80c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-80c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </button>
  </section>
</DocumentFragment>
`;

exports[`ControlPanel test cases ControlPanel renders on pause 2`] = `
<DocumentFragment>
  <section
    class="css-18a9xii"
  >
    <div
      class="css-0"
    >
      <button
        class="css-148csyh"
        data-testid="undo"
      >
        <span
          aria-label="step-backward"
          class="anticon anticon-step-backward"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="step-backward"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M347.6 528.95l383.2 301.02c14.25 11.2 35.2 1.1 35.2-16.95V210.97c0-18.05-20.95-28.14-35.2-16.94L347.6 495.05a21.53 21.53 0 000 33.9M330 864h-64a8 8 0 01-8-8V168a8 8 0 018-8h64a8 8 0 018 8v688a8 8 0 01-8 8"
            />
          </svg>
        </span>
      </button>
      <button
        class="css-2ly8m7"
      >
        <span
          aria-label="pause"
          class="anticon anticon-pause"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="pause"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M304 176h80v672h-80zm408 0h-64c-4.4 0-8 3.6-8 8v656c0 4.4 3.6 8 8 8h64c4.4 0 8-3.6 8-8V184c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </span>
      </button>
      <button
        class="css-148csyh"
        data-testid="redo"
      >
        <span
          aria-label="step-forward"
          class="anticon anticon-step-forward"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="step-forward"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M676.4 528.95L293.2 829.97c-14.25 11.2-35.2 1.1-35.2-16.95V210.97c0-18.05 20.95-28.14 35.2-16.94l383.2 301.02a21.53 21.53 0 010 33.9M694 864h64a8 8 0 008-8V168a8 8 0 00-8-8h-64a8 8 0 00-8 8v688a8 8 0 008 8"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="css-0"
    >
      <label
        class="css-jm2r67"
      >
        Rule: 
      </label>
      <select
        class="css-vqx920"
        data-testid="select"
        disabled=""
      >
        <option
          class="css-0"
          value="Hash"
        >
          Hash
        </option>
        <option
          class="css-0"
          value="Demons"
        >
          Demons
        </option>
        <option
          class="css-0"
          value="Venus"
        >
          Venus
        </option>
      </select>
    </div>
    <button
      class="css-2ly8m7"
    >
      <span
        aria-label="bg-colors"
        class="anticon anticon-bg-colors"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="bg-colors"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M766.4 744.3c43.7 0 79.4-36.2 79.4-80.5 0-53.5-79.4-140.8-79.4-140.8S687 610.3 687 663.8c0 44.3 35.7 80.5 79.4 80.5zm-377.1-44.1c7.1 7.1 18.6 7.1 25.6 0l256.1-256c7.1-7.1 7.1-18.6 0-25.6l-256-256c-.6-.6-1.3-1.2-2-1.7l-78.2-78.2a9.11 9.11 0 00-12.8 0l-48 48a9.11 9.11 0 000 12.8l67.2 67.2-207.8 207.9c-7.1 7.1-7.1 18.6 0 25.6l255.9 256zm12.9-448.6l178.9 178.9H223.4l178.8-178.9zM904 816H120c-4.4 0-8 3.6-8 8v80c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-80c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </button>
  </section>
</DocumentFragment>
`;
